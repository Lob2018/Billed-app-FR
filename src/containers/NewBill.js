import { ROUTES_PATH } from '../constants/routes.js'
import Logout from "./Logout.js"

export default class NewBill {
    constructor({ document, onNavigate, firestore, localStorage }) {
        this.document = document
        this.onNavigate = onNavigate
        this.firestore = firestore

        const formNewBill = this.document.querySelector(`form[data-testid="form-new-bill"]`)
        formNewBill.addEventListener("submit", this.handleSubmit)
        const file = this.document.querySelector(`input[data-testid="file"]`)
        file.addEventListener("change", this.handleChangeFile)
        this.fileUrl = null
        this.fileName = null
        new Logout({ document, localStorage, onNavigate })
    }
    handleChangeFile = e => {
        //file = ;
        const filePath = e.target.files[0].name.split(/\\/g)
        this.fileName = filePath[filePath.length - 1]

        const imageRegex = /\.(jpe?g|png)$/;

        const inputFileElement = this.document.querySelector(`input[data-testid="file"]`);

        if (imageRegex.test(this.fileName)) {
            inputFileElement.classList.remove('is-invalid');

        } else {
            inputFileElement.classList.add('is-invalid');
            inputFileElement.value = null;
        }
    }

    handleSubmit = e => {
        e.preventDefault()
        this.firestore.storage
            .ref(`justificatifs/${this.fileName}`)
            .put(this.document.querySelector(`input[data-testid="file"]`).files[0])
            .then(snapshot => snapshot.ref.getDownloadURL())
            .then(url => {
                this.fileUrl = url
            }).then(() => {
                console.log('e.target.querySelector(`input[data-testid="datepicker"]`).value', e.target.querySelector(`input[data-testid="datepicker"]`).value)
                const email = JSON.parse(localStorage.getItem("user")).email
                const bill = {
                    email,
                    type: e.target.querySelector(`select[data-testid="expense-type"]`).value,
                    name: e.target.querySelector(`input[data-testid="expense-name"]`).value,
                    amount: parseInt(e.target.querySelector(`input[data-testid="amount"]`).value),
                    date: e.target.querySelector(`input[data-testid="datepicker"]`).value,
                    vat: parseInt(e.target.querySelector(`input[data-testid="vat"]`).value),
                    pct: parseInt(e.target.querySelector(`input[data-testid="pct"]`).value) || 20,
                    commentary: e.target.querySelector(`textarea[data-testid="commentary"]`).value,
                    fileUrl: this.fileUrl,
                    fileName: this.fileName,
                    status: 'pending'
                }
                this.createBill(bill)
                this.onNavigate(ROUTES_PATH['Bills'])
            })
    }

    // not need to cover this function by tests
    /* istanbul ignore next */
    createBill = (bill) => {
        if (this.firestore) {
            this.firestore
                .bills()
                .add(bill)
                .then(() => {
                    this.onNavigate(ROUTES_PATH['Bills'])
                })
                .catch(error => error)
        }
    }
}