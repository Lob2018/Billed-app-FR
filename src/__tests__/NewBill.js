import { fireEvent, screen, waitFor } from "@testing-library/dom"
import NewBillUI from "../views/NewBillUI.js"
import NewBill from "../containers/NewBill.js"

import userEvent from "@testing-library/user-event";
import { localStorageMock } from "../__mocks__/localStorage";
import firebase from "../__mocks__/firebase.js";


describe("Given I am connected as an employee", () => {

    // #3 composant container/NewBill
    describe("When I am on NewBill Page", () => {
        test("Then the new bill's form should be loaded with its fields", () => {
            const html = NewBillUI()
            document.body.innerHTML = html
            expect(screen.getByTestId("form-new-bill")).toBeTruthy();
            expect(screen.getByTestId("expense-type")).toBeTruthy();
            expect(screen.getByTestId("expense-name")).toBeTruthy();
            expect(screen.getByTestId("datepicker")).toBeTruthy();
            expect(screen.getByTestId("amount")).toBeTruthy();
            expect(screen.getByTestId("vat")).toBeTruthy();
            expect(screen.getByTestId("pct")).toBeTruthy();
            expect(screen.getByTestId("commentary")).toBeTruthy();
            expect(screen.getByTestId("file")).toBeTruthy();
            expect(screen.getByRole("button")).toBeTruthy();
        })
        describe("When I upload an image in file input", () => {
            test("Then one file should be uploaded", async() => {

                document.body.innerHTML = NewBillUI();
                Object.defineProperty(window, 'localStorage', { value: localStorageMock })

                Object.defineProperty(window, 'localStorage', { value: localStorageMock })
                window.localStorage.setItem('user', JSON.stringify({
                    type: 'Employee',
                    email: "azerty@email.com",
                }))

                const newBill = new NewBill({
                    document,
                    onNavigate: () => {},
                    firestore: null,
                    localStorage: window.localStorage
                })

                const changeFile = jest.fn((e) => newBill.handleChangeFile(e))
                const file = new File(['test.jpg'], 'test.jpg', { type: 'image/jpg' })

                const input = screen.getByTestId("file")
                input.addEventListener("change", changeFile);

                userEvent.upload(input, file)

                expect(changeFile).toHaveBeenCalled();

                expect(input.files[0]).toStrictEqual(file)
                expect(input.files.item(0)).toStrictEqual(file)
                expect(input.files).toHaveLength(1)
            })
        })
        describe("When I submit the form completed", () => {
            test("Then the bill is created", async() => {

                document.body.innerHTML = NewBillUI();
                Object.defineProperty(window, 'localStorage', { value: localStorageMock })

                Object.defineProperty(window, 'localStorage', { value: localStorageMock })
                window.localStorage.setItem('user', JSON.stringify({
                    type: 'Employee',
                    email: "azerty@email.com",
                }))

                const newBill = new NewBill({
                    document,
                    onNavigate: () => {},
                    firestore: null,
                    localStorage: window.localStorage
                })

                const validBill = {
                    type: "Equipement et matériel",
                    name: "Clavier-test",
                    date: "2021-10-20",
                    amount: 10,
                    vat: 10,
                    pct: 10,
                    commentary: "Test",
                    fileUrl: "../img/0.jpg",
                    fileName: "test.jpg",
                    status: "pending"
                };

                // Load the values in fields
                screen.getByTestId("expense-type").value = validBill.type;
                screen.getByTestId("expense-name").value = validBill.name;
                // ISO 8601
                screen.getByTestId("datepicker").value = validBill.date;
                screen.getByTestId("amount").value = validBill.amount;
                screen.getByTestId("vat").value = validBill.vat;
                screen.getByTestId("pct").value = validBill.pct;
                screen.getByTestId("commentary").value = validBill.commentary;

                newBill.fileName = validBill.fileName
                newBill.fileUrl = validBill.fileUrl;

                newBill.createBill = jest.fn();
                const handleSubmit = jest.fn((e) => newBill.handleSubmit(e))

                const form = screen.getByTestId("form-new-bill");
                form.addEventListener("submit", handleSubmit);
                expect(screen.getByText('Envoyer').type).toBe('submit')

                userEvent.click(screen.getByText("Envoyer"))

                expect(handleSubmit).toHaveBeenCalled()
                expect(newBill.createBill).toHaveBeenCalledWith({
                    ...validBill
                })
            })
        })

        // #3 composant container/NewBill POST new bill
        describe("When I submit the bill's form", () => {
            test("POST bill from mock API with success", async() => {
                const postBill = jest.spyOn(firebase, "post")
                const validBill = {
                    type: "Equipement et matériel",
                    name: "Clavier-test",
                    date: "2021-10-20",
                    amount: 10,
                    vat: 10,
                    pct: 10,
                    commentary: "Test",
                    fileUrl: "https://en.wikipedia.org/wiki/File:Chrome-crash.png",
                    fileName: "logo.png",
                    email: "azerty@email.com",
                    status: "pending"
                };
                const bills = await firebase.post(validBill);
                expect(postBill).toBeCalled();
                expect(bills.data.length).toBe(5);
            })
        })
    })
})