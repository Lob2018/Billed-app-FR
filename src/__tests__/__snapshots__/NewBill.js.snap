// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given I am connected as an employee When I am on NewBill Page When I submit the form completed Then the bill is created: data url in the image-preview src for this string: "(⌐□_□)" 1`] = `undefined`;
